docs/codegen-tour.md

```diff
@@ -0,0 +1,843 @@
+we generate a 'marker' struct, which defines the type of the table, and then we
+combine it with the data via a `TableRef` struct.
+
+The `TableRef` struct looks like this:
+
+```rust
+/// Typed access to raw table data.
+pub struct TableRef<'a, T> {
+    shape: T,
+    data: FontData<'a>,
+}
+```
+
+And the definition of the table above, using a marker type, would look something
+like:
+
+```rust
+/// A marker type
+pub struct MyTableMarker;
+
+/// Instead of generating a struct for each table, we define a type alias
+pub type MyTable<'a> = TableRef<'a, MyTableMarker>;
+
+impl MyTableMarker {
+    fn format_byte_range(&self) -> Range<usize> {
+        0..u16::RAW_BYTE_LEN
+    }
+}
+
+impl MyTable<'_> {
+    fn format(&self) -> u16 {
+        let range = self.shape.format_byte_range();
+        self.data.read_at(range.start)
+    }
+}
+```
+
+To the user these two API are equivalent (you have a type `MyTable`, on which
+you can call methods to read fields) but the 'marker' pattern potentially allows
+for us to do some fancy things in the future (involving various cases where we
+want to store a type separate from a lifetime).
+
+> ***note:***
+>
+> there are also downsides of the marker pattern; in particular, currently
+> the code we generate will only compile if it is part of the `read-fonts` crate
+> itself. This isn't a major limitation, except that it makes certain kinds of
+> testing harder to do, since we can't do fancy things like generate code that
+> treated as a separate compilation unit, e.g. for use with the [`trybuild`][]
+crate.
+
+#### <a id="font-read-args"></a> `FontRead` & `FontReadWithArgs`
+
+After generating the type definitions, the next thing we generate is an
+implementation of one of [`FontRead`][] or [`FontReadWithArgs`][]. The
+`FontRead` trait is used if a table is self-describing: that is, if the data in
+the table can be fully interpreted without any external information. In some
+cases, however, this is not possible. A simple example is the [`loca` table][loca-spec]:
+the data for this table cannot be interpreted correctly without knowing the
+number of glyphs in the font (stored in the `maxp` table) as well as whether the
+format is long or short, which is stored in the `head` table.
+
+In either case, the generated table code is very similar.
+
+For the purpose of illustration, let's imagine we have a table that looks like
+this:
+
+```rust
+table Foob {
+    #[version]
+    version: BigEndian<u16>,
+    some_val: BigEndian<u32>,
+    other_val: BigEndian<u32>,
+    flags_count: BigEndian<u16>,
+    #[count($flags_count)]
+    flags: [BigEndian<u16>],
+    #[available(1)]
+    versioned_value: BigEndian<u32>,
+}
+```
+
+This generates the following code:
+
+```rust
+impl<'a> FontRead<'a> for Foob<'a> {
+    fn read(data: FontData<'a>) -> Result<Self, ReadError> {
+        let mut cursor = data.cursor();
+        let version: u16 = cursor.read()?;
```

Comment:
Suggest annotation, such as to describe

1. Which fields we actually read, e.g. always version if there is a version
1. Why we don't coalesce consecutive fixed size advances
1. The relationship to HarfBuzz sanitize, which for at least some readers will be very helpful

Reply:
for the first two points, are they answered by the section below? I walk through this line by line, explaining what code we generate for each field.

👍 on sanitize